File: desktop/ProtonEcuToolkit.Core/Kwp/PidReader.cs
using ProtonEcuToolkit.Core.Elm;
using ProtonEcuToolkit.Core.Models;

namespace ProtonEcuToolkit.Core.Kwp;

/// <summary>Sends a known PID request, parses the response, and applies its formula.</summary>
public sealed class PidReader
{
    private const int DefaultTimeoutMs = 1000;

    private readonly ElmClient _elm;
    private readonly int _timeoutMs;

    public PidReader(ElmClient elm, int timeoutMs = DefaultTimeoutMs)
    {
        _elm = elm;
        _timeoutMs = timeoutMs;
    }

    public async Task<PidReading> ReadAsync(PidDefinition pid)
    {
        var timestamp = DateTimeOffset.UtcNow;
        string? rawHex = null;

        try
        {
            rawHex = await _elm.SendCommandAsync(Protocol.BuildReadByCidRequest(pid.Id), _timeoutMs)
                .ConfigureAwait(false);
            var parsed = Protocol.ParseReadByCidResponse(rawHex, pid.Id);

            if (parsed is null)
            {
                return Failure(pid, rawHex, timestamp, $"Unrecognized response: \"{rawHex}\"");
            }
            if (!parsed.Positive)
            {
                var nrc = parsed.Nrc is { } nrcValue ? $"0x{nrcValue:x}" : "unknown";
                return Failure(pid, rawHex, timestamp, $"Negative response, NRC={nrc}");
            }
            if (parsed.DataBytes.Length < pid.ByteLength)
            {
                return Failure(
                    pid, rawHex, timestamp,
                    $"Expected {pid.ByteLength} data byte(s), got {parsed.DataBytes.Length}");
            }

            return new PidReading(pid.Id, pid.Name, pid.Unit, pid.Decode(parsed.DataBytes), rawHex, timestamp);
        }
        catch (Exception err)
        {
            return Failure(pid, rawHex, timestamp, err.Message);
        }
    }

    public async Task<List<PidReading>> ReadAllAsync(IReadOnlyList<PidDefinition> pids)
    {
        var readings = new List<PidReading>(pids.Count);
        foreach (var pid in pids)
        {
            readings.Add(await ReadAsync(pid).ConfigureAwait(false));
        }
        return readings;
    }

    private static PidReading Failure(PidDefinition pid, string? rawHex, DateTimeOffset timestamp, string error) =>
        new(pid.Id, pid.Name, pid.Unit, null, rawHex, timestamp, error);
}

File: src/server/kwp/PidReader.ts
import { ElmClient } from "../elm/ElmClient.js";
import type { PidDefinition, PidReading } from "../../shared/types.js";
import { buildReadByCidRequest, parseReadByCidResponse } from "./protocol.js";

const DEFAULT_TIMEOUT_MS = 1000;

/** Sends a known PID request, parses the response, and applies its formula. */
export class PidReader {
  constructor(
    private readonly elm: ElmClient,
    private readonly timeoutMs = DEFAULT_TIMEOUT_MS,
  ) {}

  async read(pid: PidDefinition): Promise<PidReading> {
    const timestamp = Date.now();
    let rawHex: string | null = null;

    try {
      rawHex = await this.elm.sendCommand(buildReadByCidRequest(pid.id), {
        timeoutMs: this.timeoutMs,
      });
      const parsed = parseReadByCidResponse(rawHex, pid.id);

      if (!parsed) {
        return this.failure(pid, rawHex, timestamp, `Unrecognized response: "${rawHex}"`);
      }
      if (!parsed.positive) {
        const nrc = parsed.nrc !== undefined ? `0x${parsed.nrc.toString(16)}` : "unknown";
        return this.failure(pid, rawHex, timestamp, `Negative response, NRC=${nrc}`);
      }
      if (parsed.dataBytes.length < pid.byteLength) {
        return this.failure(
          pid,
          rawHex,
          timestamp,
          `Expected ${pid.byteLength} data byte(s), got ${parsed.dataBytes.length}`,
        );
      }

      return {
        id: pid.id,
        name: pid.name,
        unit: pid.unit,
        value: pid.decode(parsed.dataBytes),
        rawHex,
        timestamp,
      };
    } catch (err) {
      return this.failure(pid, rawHex, timestamp, err instanceof Error ? err.message : String(err));
    }
  }

  async readAll(pids: PidDefinition[]): Promise<PidReading[]> {
    const readings: PidReading[] = [];
    for (const pid of pids) {
      readings.push(await this.read(pid));
    }
    return readings;
  }

  private failure(pid: PidDefinition, rawHex: string | null, timestamp: number, error: string): PidReading {
    return { id: pid.id, name: pid.name, unit: pid.unit, value: null, rawHex, timestamp, error };
  }
}
